File: problems/15_greedy/0055_jump_game/solution.py
"""
Problem: 0055. Jump Game
"""
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        reach = 0
        for i, x in enumerate(nums):
            if i > reach:
                return False
            if i + x > reach:
                reach = i + x
        return True


if __name__ == "__main__":
    sol = Solution()
    assert sol.canJump([2, 3, 1, 1, 4]) is True
    assert sol.canJump([3, 2, 1, 0, 4]) is False
    assert sol.canJump([0]) is True
    assert sol.canJump([2, 0]) is True
    assert sol.canJump([1, 0, 1]) is False
    assert sol.canJump([2, 0, 0]) is True
    print("All tests passed!")

File: problems/15_greedy/0055_jump_game/solution.dart
// 0055. Jump Game
bool canJump(List<int> nums) {
  var reach = 0;
  for (var i = 0; i < nums.length; i++) {
    if (i > reach) return false;
    if (i + nums[i] > reach) reach = i + nums[i];
  }
  return true;
}
void main() {
  if (canJump([2, 3, 1, 1, 4]) != true) throw 'a';
  if (canJump([3, 2, 1, 0, 4]) != false) throw 'b';
  if (canJump([0]) != true) throw 'c';
  if (canJump([2, 0]) != true) throw 'd';
  if (canJump([1, 0, 1]) != false) throw 'e';
  if (canJump([2, 0, 0]) != true) throw 'f';
  print('All tests passed!');
}
